<!DOCTYPE html>
<html lang="en"><head>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="presentation_files/libs/clipboard/clipboard.min.js"></script>
<script src="presentation_files/libs/quarto-html/tabby.min.js"></script>
<script src="presentation_files/libs/quarto-html/popper.min.js"></script>
<script src="presentation_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="presentation_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="presentation_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="presentation_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.4.549">

  <meta name="author" content="Connor M. French">
  <title>The influence of environmental change on genetic diversity across spatial and taxonomic scales</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="presentation_files/libs/revealjs/dist/theme/quarto.css">
  <link href="presentation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="presentation_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
  
  <script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">The influence of environmental change on genetic diversity across spatial and taxonomic scales</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Connor M. French 
</div>
</div>
</div>

</section>
<section id="what-the-audience-has-been-introduced-to" class="slide level2">
<h2>What the audience has been introduced to</h2>
<ul>
<li class="fragment"><p>What an SDM is</p></li>
<li class="fragment"><p>The idea of transforming SDM output into the parameters for a 2D stepping-stone model (will briefly review it)</p></li>
<li class="fragment"><p>what these types of models are useful for</p></li>
</ul>
</section>
<section id="spaceprime-a-python-package-to-facilitate-spatially-explicit-coalescent-models-in-msprime" class="slide level2">
<h2><em>spaceprime</em>: a Python package to facilitate spatially-explicit coalescent models in <em>msprime</em></h2>
</section>
<section id="spaceprime-makes-complex-eco-evolutionary-models-a-little-easier" class="slide level2">
<h2><em>spaceprime</em> makes complex eco-evolutionary models a little easier</h2>
<ul>
<li class="fragment"><p>The typical workflow of an integrative distributional, demographic coalescent (iDDC) model involves extensive domain knowledge <em>and</em> computational know-how</p></li>
<li class="fragment"><p><em>spaceprime</em> aims to reduce the overhead on computational know-how</p></li>
</ul>
</section>
<section class="slide level2">


<img data-src="images/conceptual-fig.png" class="quarto-figure quarto-figure-center r-stretch"><aside class="notes">
<p>iDDC models involve converting SDMs into parameters for spatial coalescent models</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section class="slide level2">


<img data-src="images/msprime_logo.svg" class="quarto-figure quarto-figure-center r-stretch" style="width:20cm"><p class="caption">“A population genetics simulator of ancestry and DNA sequence evolution”</p><aside class="notes">
<p>the parameters are passed to msprime to simulate the ancestry and genetic diversity of user-provided samples</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section class="slide level2">

<div class="v-center-container">
<p><em>But first, a simple msprime example</em></p>
</div>
</section>
<section class="slide level2">


<img data-src="presentation_files/figure-revealjs/msprime-ex-plot-output-1.svg" id="msprime-ex-plot" class="r-stretch"><aside class="notes">
<p>To impress upon you why an intermediate tool may be needed, here is an example msprime simulation for a simple demographic scenario. Three populations are initialized with their own sizes- 10,000, 5,000, and 1,000 individuals respectively. The A and B populations merge into C 1000 generations ago and migration is continuous and symmetric between populations A and B.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section class="slide level2">

<div id="msprime-ex-code" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href=""></a><span class="im">import</span> msprime</span>
<span id="cb1-2"><a href=""></a></span>
<span id="cb1-3"><a href=""></a>demography <span class="op">=</span> msprime.Demography()</span>
<span id="cb1-4"><a href=""></a>demography.add_population(name<span class="op">=</span><span class="st">"A"</span>, initial_size<span class="op">=</span><span class="dv">10_000</span>)</span>
<span id="cb1-5"><a href=""></a>demography.add_population(name<span class="op">=</span><span class="st">"B"</span>, initial_size<span class="op">=</span><span class="dv">5_000</span>)</span>
<span id="cb1-6"><a href=""></a>demography.add_population(name<span class="op">=</span><span class="st">"C"</span>, initial_size<span class="op">=</span><span class="dv">1_000</span>)</span>
<span id="cb1-7"><a href=""></a>demography.add_population_split(time<span class="op">=</span><span class="dv">1000</span>, derived<span class="op">=</span>[<span class="st">"A"</span>, <span class="st">"B"</span>], ancestral<span class="op">=</span><span class="st">"C"</span>)</span>
<span id="cb1-8"><a href=""></a>demography.set_symmetric_migration_rate(populations<span class="op">=</span>[<span class="st">"A"</span>, <span class="st">"B"</span>], rate<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb1-9"><a href=""></a>ts <span class="op">=</span> msprime.sim_ancestry(samples<span class="op">=</span>{<span class="st">"A"</span>: <span class="dv">2</span>, <span class="st">"B"</span>: <span class="dv">2</span>}, demography<span class="op">=</span>demography, random_seed<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb1-10"><a href=""></a>ts <span class="op">=</span> msprime.sim_mutations(ts, rate<span class="op">=</span><span class="fl">3e-4</span>, random_seed <span class="op">=</span> <span class="dv">42</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="presentation_files/figure-revealjs/msprime-ex-draw-output-1.svg" id="msprime-ex-draw" class="quarto-figure quarto-figure-center r-stretch"><aside class="notes">
<p>The ancestry of four individuals- 2 from A and 2 from B- are simulated and mutations are overlaid on the tree sequence. The little tree represents the ancestral relationships among the sampled chromosomes- two per individual. The red x’s are mutations that have occurred at specific points in the past.</p>
<p>Try to do this for thousands of populations that are all connected by migration and change over time, and the task gets much more difficult. Both to code and to figure out if your code makes sense. In addition, if you are trying to compare this output to a typical empirical data set, you need to convert the tree sequence into a more typical empirical data style, like a genotype matrix, that you then calculate summary statistics from. It’s a lot!</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="spaceprime-makes-the-journey-from-sdm-to-analysis-easier" class="slide level2">
<h2><em>spaceprime</em> makes the journey from SDM to analysis easier</h2>
<div class="columns">
<div class="column" style="width:50%;">
<h4 id="simulation">Simulation</h4>
<ul>
<li class="fragment"><code>spaceprime.demography</code></li>
<li class="fragment"><code>msprime.sim_*</code></li>
</ul>
<h4 id="analysis">Analysis</h4>
<ul>
<li class="fragment"><code>spaceprime.analysis</code></li>
</ul>
</div><div class="column" style="width:50%;">
<h4 id="debugging-and-utilities">Debugging and utilities</h4>
<ul>
<li class="fragment"><code>spaceprime.plotting</code></li>
<li class="fragment"><code>spaceprime.utilities</code></li>
</ul>
<h4 id="command-line">Command line</h4>
<ul>
<li class="fragment"><code>spaceprime –-params</code></li>
</ul>
</div>
</div>
<aside class="notes">
<p>The package is broken up into four main modules: <code>demography</code>, <code>analysis</code>, <code>plotting</code>, and <code>utilities</code>. The <code>demography</code> module is for setting up the spatially-explicit coalescent model, the <code>analysis</code> module is for calculating summary statistics from the output of the simulations, the <code>plotting</code> module is for visualizing the demographic model for debugging purposes, and the <code>utilities</code> module is for debugging and other miscellaneous tasks. The package also has a command line interface that allows users to run batches of simulations in parallel, using an annotated params file to specify input parameters. Walking you through all of the functions in the package would take a long time, so I will focus on a high-level overview of how everything fits together.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="simulating-the-demographic-history-of-the-leaf-frog-phyllomedusa-distincta" class="slide level2 nostretch">
<h2>Simulating the demographic history of the leaf frog <em>Phyllomedusa distincta</em></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="images/pdistincta.jpg" style="height:10cm"><br>
<img data-src="images/habitat.jpg" style="height:10cm"></p>
</div><div class="column" style="width:50%;">
<div class="r-stack">
<div class="fragment fade-in-then-out">
<div id="cell-pdistincta-map" class="cell" data-execution_count="7">
<div class="cell-output cell-output-display">
<div>
<figure>
<p><img data-src="presentation_files/figure-revealjs/pdistincta-map-output-1.svg" id="pdistincta-map"></p>
</figure>
</div>
</div>
</div>
</div>
<p><img data-src="images/stability_animation.gif" class="fragment fade-in"></p>
</div>
</div>
</div>
<aside class="notes">
<p>I’ll walk through the main features of <em>spaceprime</em> by going through an example analysis of the leaf frog <em>Phyllomedusa distincta</em>, which is distributed across the central and southern Brazilian Atlantic Forest.</p>
<p>Funny enough, this is the only chapter I did field work for!</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="simulation-1" class="slide level2">
<h2>Simulation</h2>

<img data-src="images/simulation.svg" class="quarto-figure quarto-figure-center r-stretch"><div class="fragment">
<div id="pdistincta-sim" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href=""></a><span class="im">import</span> spaceprime <span class="im">as</span> sp</span>
<span id="cb2-2"><a href=""></a><span class="im">import</span> geopandas <span class="im">as</span> gpd</span>
<span id="cb2-3"><a href=""></a><span class="im">import</span> msprime</span>
<span id="cb2-4"><a href=""></a></span>
<span id="cb2-5"><a href=""></a>projections <span class="op">=</span> rasterio.<span class="bu">open</span>(<span class="st">"data/projections_agg.tif"</span>)</span>
<span id="cb2-6"><a href=""></a>locs <span class="op">=</span> gpd.read_file(<span class="st">"data/distincta_localities_ex.geojson"</span>)</span>
<span id="cb2-7"><a href=""></a>mpt <span class="op">=</span> sp.mtp_thresh_from_coords(projections, locs)</span>
<span id="cb2-8"><a href=""></a>demes <span class="op">=</span> sp.raster_to_demes(projections, transformation<span class="op">=</span><span class="st">"linear"</span>, max_local_size<span class="op">=</span><span class="dv">1000</span>, threshold<span class="op">=</span>mpt)</span>
<span id="cb2-9"><a href=""></a>d <span class="op">=</span> sp.spDemography()</span>
<span id="cb2-10"><a href=""></a>d.stepping_stone_2d(demes, rate<span class="op">=</span><span class="fl">0.001</span>, scale<span class="op">=</span><span class="va">True</span>, timesteps<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb2-11"><a href=""></a>d.add_ancestral_populations(anc_sizes <span class="op">=</span> [<span class="dv">100000</span>], merge_time <span class="op">=</span> <span class="dv">23000</span>)</span>
<span id="cb2-12"><a href=""></a>sample_dict, _, _ <span class="op">=</span> sp.coords_to_sample_dict(projections, locs)</span>
<span id="cb2-13"><a href=""></a>ts <span class="op">=</span> msprime.sim_ancestry(demography<span class="op">=</span>d, samples<span class="op">=</span>sample_dict, sequence_length<span class="op">=</span><span class="fl">1e6</span>, recombination_rate<span class="op">=</span><span class="fl">1e-9</span>, random_seed<span class="op">=</span><span class="dv">43</span>, record_provenance<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb2-14"><a href=""></a>ts <span class="op">=</span> msprime.sim_mutations(ts, rate<span class="op">=</span><span class="fl">1e-8</span>, random_seed<span class="op">=</span><span class="dv">99</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="simulate-massive-models-with-a-minimal-amount-of-code" class="slide level2">
<h2>Simulate massive models with a minimal amount of code</h2>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="simulate-174446-demographic-events">Simulate 174,446 demographic events</h3>
<div id="pdistincta-sim2" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href=""></a><span class="im">import</span> spaceprime <span class="im">as</span> sp</span>
<span id="cb3-2"><a href=""></a><span class="im">import</span> geopandas <span class="im">as</span> gpd</span>
<span id="cb3-3"><a href=""></a><span class="im">import</span> msprime</span>
<span id="cb3-4"><a href=""></a></span>
<span id="cb3-5"><a href=""></a>projections <span class="op">=</span> rasterio.<span class="bu">open</span>(<span class="st">"data/projections_agg.tif"</span>)</span>
<span id="cb3-6"><a href=""></a>locs <span class="op">=</span> gpd.read_file(<span class="st">"data/distincta_localities_ex.geojson"</span>)</span>
<span id="cb3-7"><a href=""></a>mpt <span class="op">=</span> sp.mtp_thresh_from_coords(projections, locs)</span>
<span id="cb3-8"><a href=""></a>demes <span class="op">=</span> sp.raster_to_demes(projections, transformation<span class="op">=</span><span class="st">"linear"</span>, max_local_size<span class="op">=</span><span class="dv">1000</span>, threshold<span class="op">=</span>mpt)</span>
<span id="cb3-9"><a href=""></a>d <span class="op">=</span> sp.spDemography()</span>
<span id="cb3-10"><a href=""></a>d.stepping_stone_2d(demes, rate<span class="op">=</span><span class="fl">0.001</span>, scale<span class="op">=</span><span class="va">True</span>, timesteps<span class="op">=</span><span class="dv">1000</span>)</span>
<span id="cb3-11"><a href=""></a>d.add_ancestral_populations(anc_sizes <span class="op">=</span> [<span class="dv">100000</span>], merge_time <span class="op">=</span> <span class="dv">23000</span>)</span>
<span id="cb3-12"><a href=""></a>sample_dict, _, _ <span class="op">=</span> sp.coords_to_sample_dict(projections, locs)</span>
<span id="cb3-13"><a href=""></a>ts <span class="op">=</span> msprime.sim_ancestry(demography<span class="op">=</span>d, samples<span class="op">=</span>sample_dict, sequence_length<span class="op">=</span><span class="fl">1e6</span>, recombination_rate<span class="op">=</span><span class="fl">1e-9</span>, random_seed<span class="op">=</span><span class="dv">43</span>, record_provenance<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb3-14"><a href=""></a>ts <span class="op">=</span> msprime.sim_mutations(ts, rate<span class="op">=</span><span class="fl">1e-8</span>, random_seed<span class="op">=</span><span class="dv">99</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:50%;">
<h3 id="simulate-4-demographic-events">Simulate 4 demographic events</h3>
<div id="msprime-ex-code2" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href=""></a><span class="im">import</span> msprime</span>
<span id="cb4-2"><a href=""></a></span>
<span id="cb4-3"><a href=""></a>demography <span class="op">=</span> msprime.Demography()</span>
<span id="cb4-4"><a href=""></a>demography.add_population(name<span class="op">=</span><span class="st">"A"</span>, initial_size<span class="op">=</span><span class="dv">10_000</span>)</span>
<span id="cb4-5"><a href=""></a>demography.add_population(name<span class="op">=</span><span class="st">"B"</span>, initial_size<span class="op">=</span><span class="dv">5_000</span>)</span>
<span id="cb4-6"><a href=""></a>demography.add_population(name<span class="op">=</span><span class="st">"C"</span>, initial_size<span class="op">=</span><span class="dv">1_000</span>)</span>
<span id="cb4-7"><a href=""></a>demography.add_population_split(time<span class="op">=</span><span class="dv">1000</span>, derived<span class="op">=</span>[<span class="st">"A"</span>, <span class="st">"B"</span>], ancestral<span class="op">=</span><span class="st">"C"</span>)</span>
<span id="cb4-8"><a href=""></a>demography.set_symmetric_migration_rate(populations<span class="op">=</span>[<span class="st">"A"</span>, <span class="st">"B"</span>], rate<span class="op">=</span><span class="fl">0.1</span>)</span>
<span id="cb4-9"><a href=""></a>ts <span class="op">=</span> msprime.sim_ancestry(samples<span class="op">=</span>{<span class="st">"A"</span>: <span class="dv">2</span>, <span class="st">"B"</span>: <span class="dv">2</span>}, demography<span class="op">=</span>demography, random_seed<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb4-10"><a href=""></a>ts <span class="op">=</span> msprime.sim_mutations(ts, rate<span class="op">=</span><span class="fl">3e-4</span>, random_seed <span class="op">=</span> <span class="dv">42</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="analyze" class="slide level2">
<h2>Analyze</h2>

<img data-src="images/analysis.svg" class="quarto-figure quarto-figure-center r-stretch"></section>
<section class="slide level2">

<div class="columns">
<div class="column" style="width:50%;">
<h3 id="convert-your-data-to-a-genotype-table-and-filter">Convert your data to a genotype table and filter</h3>
<div id="cell-genomat" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href=""></a>gt <span class="op">=</span> ts.genotype_matrix()</span>
<span id="cb5-2"><a href=""></a></span>
<span id="cb5-3"><a href=""></a>gt_filt <span class="op">=</span> sp.filter_gt(gt, missing_data_perc <span class="op">=</span> <span class="fl">0.4</span>, r2_thresh <span class="op">=</span> <span class="fl">0.1</span>)</span>
<span id="cb5-4"><a href=""></a></span>
<span id="cb5-5"><a href=""></a>gt_filt[<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="genomat" class="cell-output cell-output-display" data-execution_count="42">
<div class="allel allel-DisplayAs2D"><span>&lt;GenotypeArray shape=(11018, 172, 2) dtype=int32&gt;</span>
<table data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: center;" data-quarto-table-cell-role="th"></th>
<th style="text-align: center;" data-quarto-table-cell-role="th">0</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">1</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">2</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">3</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">4</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">...</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">167</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">168</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">169</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">170</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">171</th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">1/0</td>
<td style="text-align: center;">...</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">1</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">...</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td></td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">2</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">...</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">...</td>
<td colspan="12" style="text-align: center;">...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">11015</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">...</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">11016</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/1</td>
<td style="text-align: center;">1/1</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">...</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td></td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" style="text-align: center; background-color: white; border-right: 1px solid black;">11017</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">...</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td style="text-align: center;">0/0</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment fade-in">
<!-- -->
<h3 id="calculate-summary-statistics-to-compare-with-empirical-data">Calculate summary statistics to compare with empirical data</h3>
<div id="cell-sumstats" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href=""></a>coords_dict <span class="op">=</span> sp.coords_to_deme_dict(projections, locs)</span>
<span id="cb6-2"><a href=""></a>ss <span class="op">=</span> sp.calc_sumstats(gt_filt[<span class="dv">1</span>], coords_dict, return_df <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb6-3"><a href=""></a></span>
<span id="cb6-4"><a href=""></a>ss</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="sumstats" class="cell-output cell-output-display" data-execution_count="44">
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>

<table class="dataframe" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header" style="text-align: right;">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sfs_h1</th>
<th data-quarto-table-cell-role="th">sfs_h2</th>
<th data-quarto-table-cell-role="th">pi</th>
<th data-quarto-table-cell-role="th">pi_sd</th>
<th data-quarto-table-cell-role="th">taj_d</th>
<th data-quarto-table-cell-role="th">taj_d_sd</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.085726</td>
<td>0.037712</td>
<td>0.052481</td>
<td>0.086362</td>
<td>-1.97303</td>
<td>0.105494</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="throw-it-all-together-on-the-command-line" class="slide level2">
<h2>Throw it all together on the command line</h2>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="spaceprime---params"><code>spaceprime --params</code></h3>
</div><div class="column" style="width:50%;">
<div class="fragment fade-in">
<div id="params-file" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource python number-lines code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href=""></a><span class="co"># config file for the spaceprime command line tool</span></span>
<span id="cb7-2"><a href=""></a><span class="co"># This file is a template and should be modified to suit the user's needs.</span></span>
<span id="cb7-3"><a href=""></a><span class="co"># For more information on the parameters, see the documentation at https://connor-french.github.io/spaceprime/</span></span>
<span id="cb7-4"><a href=""></a></span>
<span id="cb7-5"><a href=""></a><span class="co"># All list entries must be specified like this: [entry1, entry2, entry3]</span></span>
<span id="cb7-6"><a href=""></a><span class="co"># All pairs of ints, used for specifying ranges, must be specified like this: [int1, int2]</span></span>
<span id="cb7-7"><a href=""></a><span class="co"># All paths must be specified like this: "path/to/file"</span></span>
<span id="cb7-8"><a href=""></a><span class="co"># All boolean values must be specified like this: true or false</span></span>
<span id="cb7-9"><a href=""></a></span>
<span id="cb7-10"><a href=""></a><span class="co"># global arguments #</span></span>
<span id="cb7-11"><a href=""></a>raster: <span class="st">"projections.tif"</span> <span class="co">#Path to raster file. Can be any raster format that rasterio can read.</span></span>
<span id="cb7-12"><a href=""></a>coords: <span class="st">"distincta_localities_ex.csv"</span> <span class="co">#Path to file containing sampling coordinates. Must be a CSV file containing columns 'longitude' and 'latitude'.</span></span>
<span id="cb7-13"><a href=""></a>individuals: <span class="st">"distincta_localities_ex.csv"</span> <span class="co">#Optional. Either a list or a path to file containing individual IDs. If a CSV, must contain a column named 'individual_id'. </span></span>
<span id="cb7-14"><a href=""></a>  </span>
<span id="cb7-15"><a href=""></a><span class="co"># demography setup arguments #</span></span>
<span id="cb7-16"><a href=""></a>normalize: false <span class="co">#Normalize raster values to [0,1] range. </span></span>
<span id="cb7-17"><a href=""></a>transformation: <span class="st">"linear"</span> <span class="co">#Transformation to apply to raster. Options are 'linear', 'threshold', 'sigmoid'. Default is 'linear'.</span></span>
<span id="cb7-18"><a href=""></a>max_local_size: [<span class="dv">500</span>,<span class="dv">2000</span>] <span class="co">#Maximum size of local demes. Accepts a single int or a pair of ints, specified like [100, 1000].</span></span>
<span id="cb7-19"><a href=""></a>threshold: null <span class="co">#0.0942648 #Optional. Threshold value for a thresholded transformation. Accepts a single float or a pair of floats.</span></span>
<span id="cb7-20"><a href=""></a>inflection_point: [<span class="fl">0.5</span>] <span class="co">#Inflection point value for a sigmoid transformation. Accepts a single int or a pair of ints.</span></span>
<span id="cb7-21"><a href=""></a>slope: [<span class="fl">0.05</span>] <span class="co">#Slope value for a sigmoid transformation. Accepts a single float or a pair of floats.</span></span>
<span id="cb7-22"><a href=""></a>mig_rate: [<span class="fl">0.0001</span>,<span class="fl">0.005</span>] <span class="co">#Optional. Global migration rate between demes. Accepts a single float or a pair of floats.</span></span>
<span id="cb7-23"><a href=""></a>scale: true <span class="co">#Scale the migration rate by donor and recipient deme size using the equation m = (N_donor / N_recipient) * m_global.</span></span>
<span id="cb7-24"><a href=""></a>anc_pop_id: null <span class="co">#Optional. List of ancestral population IDs or path to CSV file with column 'anc_pop_id'. Both have to be the same length as --coords. Only supply if you want to add ancestral populations.</span></span>
<span id="cb7-25"><a href=""></a>timesteps: <span class="dv">1000</span> <span class="co">#The list of timesteps representing the amount of time passing between each demographic event, in generations. If a single integer is provided, the function assumes that the time steps are equal</span></span>
<span id="cb7-26"><a href=""></a>anc_sizes: [<span class="dv">700000</span>,<span class="dv">1200000</span>] <span class="co">#Optional. List of sizes for ancestral populations. Accepts a list of ints or a list of pairs of ints.</span></span>
<span id="cb7-27"><a href=""></a>merge_time: <span class="dv">23000</span> <span class="co">#Time that demes merge into one or more ancestral populations. Measured in generations. Accepts a single int or a pair of ints.</span></span>
<span id="cb7-28"><a href=""></a>anc_merge_time: null <span class="co">#Merge time for ancestral populations. Measured in generations. Accepts a single int or a pair of ints.</span></span>
<span id="cb7-29"><a href=""></a>anc_merge_size: null <span class="co">#Merge size for ancestral populations. Accepts a single int or a pair of ints.</span></span>
<span id="cb7-30"><a href=""></a>anc_mig_rate: null <span class="co">#Migration rate between ancestral populations. Accepts a single int or a pair of ints.</span></span>
<span id="cb7-31"><a href=""></a></span>
<span id="cb7-32"><a href=""></a><span class="co"># simulation arguments #</span></span>
<span id="cb7-33"><a href=""></a>seq_length: <span class="dv">1000000</span> <span class="co">#Length of sequence to simulate. Measured in base pairs.</span></span>
<span id="cb7-34"><a href=""></a>mutation_rate: [<span class="fl">0.00000008</span>] <span class="co">#Mutation rate per base pair per generation. Accepts a single float or a pair of floats.</span></span>
<span id="cb7-35"><a href=""></a>recombination_rate: [<span class="fl">0.000000001</span>] <span class="co">#Recombination rate per base pair per generation. Accepts a single float or a pair of floats.</span></span>
<span id="cb7-36"><a href=""></a>ploidy: <span class="dv">2</span> <span class="co">#Ploidy of individuals. Accepts an int.</span></span>
<span id="cb7-37"><a href=""></a>num_param_combos: <span class="dv">100</span> <span class="co">#Number of parameter combinations to simulate. Accepts an int.</span></span>
<span id="cb7-38"><a href=""></a>num_coalescent_sims: <span class="dv">1</span> <span class="co">#Number of coalescent simulations to run per parameter combination. Accepts an int.</span></span>
<span id="cb7-39"><a href=""></a></span>
<span id="cb7-40"><a href=""></a><span class="co"># analysis arguments #</span></span>
<span id="cb7-41"><a href=""></a>missing_data_perc: <span class="dv">0</span> <span class="co">#Percentage of data masked as missing. Accepts a single float between 0 and 1.</span></span>
<span id="cb7-42"><a href=""></a>r2_thresh: <span class="fl">0.1</span> <span class="co">#Threshold value for filtering sites based on linkage disequilibrium R^2.</span></span>
<span id="cb7-43"><a href=""></a>filter_monomorphic: true <span class="co">#Filter out monomorphic sites.</span></span>
<span id="cb7-44"><a href=""></a>filter_singletons: true <span class="co">#Filter out singleton sites.</span></span>
<span id="cb7-45"><a href=""></a>sumstats: <span class="st">"all"</span> <span class="co">#List of summary statistics to calculate. Options are 'pi', 'tajima_d', 'sfs_h', 'fst', 'dxy', 'ibd', 'all'. For explanation of summary statistics, see the documentation. Default is all.</span></span>
<span id="cb7-46"><a href=""></a>within_anc_pop_sumstats: false <span class="co">#Whether to calculate summary statistics within ancestral populations.</span></span>
<span id="cb7-47"><a href=""></a>between_anc_pop_sumstats: false <span class="co">#Whether to calculate Fst and/or Dxy between ancestral populations.</span></span>
<span id="cb7-48"><a href=""></a></span>
<span id="cb7-49"><a href=""></a><span class="co"># output arguments #</span></span>
<span id="cb7-50"><a href=""></a><span class="bu">map</span>: false <span class="co">#Simulate across the full landscape and return a map of genetic diversity</span></span>
<span id="cb7-51"><a href=""></a>out_type: <span class="dv">2</span> <span class="co">#Type of output. 0 for writing tree sequences to file, 1 for writing VCFs to file, 2 for a CSV of genetic summary statistics, or 3 for writing all outputs to files.</span></span>
<span id="cb7-52"><a href=""></a>out_folder: <span class="st">"output"</span> <span class="co">#Path to output folder. Default is the current working directory.</span></span>
<span id="cb7-53"><a href=""></a>out_prefix: <span class="st">"onepop"</span> <span class="co">#Prefix for output files. Default is 'spaceprime'.</span></span>
<span id="cb7-54"><a href=""></a>verbose: false <span class="co">#Print verbose output.</span></span>
<span id="cb7-55"><a href=""></a></span>
<span id="cb7-56"><a href=""></a><span class="co"># CPU #</span></span>
<span id="cb7-57"><a href=""></a>cpu: <span class="dv">15</span> <span class="co">#Number of CPUs to use. Accepts an int.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment fade-in smalltext">
<!-- -->
<h3 id="max_local_size-5002000"><code>max_local_size: [500,2000]</code></h3>
<h3 id="cpu-15"><code>cpu: 15</code></h3>
<h3 id="map-false"><code>map: false</code></h3>
</div>
</div>
</div>
<aside class="notes">
<p>You can specify priors, which will be uniform across the range of values you specify.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="i-simulated-four-demographic-models" class="slide level2">
<h2>I simulated four demographic models</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="images/hyp1-const-one.svg" class="quarto-figure quarto-figure-center"> <img data-src="images/hyp2-const-two.svg" class="quarto-figure quarto-figure-center"></p>
</div><div class="column" style="width:50%;">
<p><img data-src="images/stability_animation.gif" class="quarto-figure quarto-figure-center"> <img data-src="images/hyp3-fluct-one.svg" class="quarto-figure quarto-figure-center"> <img data-src="images/stability_animation.gif" class="quarto-figure quarto-figure-center"> <img data-src="images/hyp4-fluct-two.svg" class="quarto-figure quarto-figure-center"></p>
</div>
</div>
</section>
<section class="slide level2">


<img data-src="images/gendiv_fig_maps.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section class="slide level2">


<img data-src="images/gendiv_fig_PCA.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="computational-resources" class="slide level2">
<h2>Computational resources</h2>

<img data-src="images/sim_time_fig.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="documentation" class="slide level2">
<h2>Documentation</h2>
<p><a href="https://connor-french.github.io/spaceprime/">https://connor-french.github.io/spaceprime/</a></p>
<aside class="notes">
<p>Quickstart, “Intro to spaceprime for R users”</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="conclusions" class="slide level2">
<h2>Conclusions</h2>
<ul>
<li class="fragment">TBD</li>
</ul>
</section>
<section id="thanks" class="slide level2">
<h2>Thanks!</h2>
<div class="quarto-auto-generated-content">
<div class="footer footer-default">

</div>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="presentation_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="presentation_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="presentation_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="presentation_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1920,

        height: 1080,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>